ballerina/variable_validator.bal

```diff
@@ -149,9 +168,87 @@ class VariableValidator {
             argument.setVariableValue(value);
         } else {
             string message = string`Variable "$${<string> argument.getVariableName()}" got invalid value ` +
-            string`${value.toString()};Expected type ${getTypeName(argument)}. ${getTypeName(argument)}` +
+            string`${value.toString()};Expected type ${varDef.kind}. ${getTypeName(argument)}` +
             string` cannot represent value: ${value.toString()}`;
-            self.errors.push(getErrorDetailRecord(message, location)); 
+            self.errors.push(getErrorDetailRecord(message, argument.getLocation()));
         }
     }
+
+    isolated function checkVariableUsageCompatibility(__InputValue[] inputValues, parser:VariableDefinition varDef,
+                                                      parser:ArgumentNode argNode) {
+        __InputValue? inputValue = getInputValueFromArray(inputValues, argNode.getName());
+        if inputValue is __InputValue {
+            if !self.isVariableUsageAllowed(varDef, inputValue) {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            }
+        }
+    }
+
+    isolated function isVariableUsageAllowed(parser:VariableDefinition varDef, __InputValue inputValue) returns boolean {
+        __Type? varType = self.getTypeRecordFromTypeName(varDef.kind);
+        if varType is __Type {
+            if inputValue.'type.kind == NON_NULL && varType.kind != NON_NULL {
+                if inputValue?.defaultValue is () && varDef?.defaultValue is () {
+                    return false;
+                }
+                return self.areTypesCompatible(varType, getOfType(inputValue.'type));
+            }
+            return self.areTypesCompatible(varType, inputValue.'type);
+        }
+        return false;
+    }
+
+    isolated function areTypesCompatible(__Type varType, __Type inputType) returns boolean {
+        if inputType.kind == NON_NULL {
+            if varType.kind != NON_NULL {
+                return false;
+            }
+            return self.areTypesCompatible(getOfType(varType), getOfType(inputType));
```

Comment:
`getOfType()` here will strip the type completely, which means removes all the wrapping types. Is that what we need here? Or do we need just `varType.ofType` ?

Reply:
This should be `varType.ofType`. Here we want to remove only the first wrapper type.